### PR TITLE
Add `StatusProto.toStatusException` overload to accept `Throwable`

### LINKED ISSUE #10900: Ability to propagate `cause` when using `StatusProto.toStatusException`
### Is your feature request related to a problem?

When converting a `com.google.rpc.Status` to an `io.grpc.StatusException` I can't find a way to propagate a `cause` with the resulting `StatusException`.

Currently `StatusProto.toStatusException` is defined as:

```java
  public static StatusException toStatusException(com.google.rpc.Status statusProto) {
    return toStatus(statusProto).asException(toMetadata(statusProto));
  }
```

However, using this method there doesn't seem to be any way of providing a `cause`.

### Describe the solution you'd like

My ideal would be to make `StatusProto.toStatus` public, which would allow creating the status and calling `asException` something callers could do themselves. This would also be helpful in scenarios where callers require a `Status` instead of a `StatusException`.

```java
  private static Status toStatus(com.google.rpc.Status statusProto) {
    Status status = Status.fromCodeValue(statusProto.getCode());
    checkArgument(status.getCode().value() == statusProto.getCode(), "invalid status code");
    return status.withDescription(statusProto.getMessage());
  }
```

### Describe alternatives you've considered

It feels a little clunky, but I'd be open to having an extra parameter on `toStatusException` for the `cause`, or another solution depending on the preferences of `grpc-java` maintainers.

I've also tried calling `initCause` on the resulting StatusException, but that isn't possible because the cause is set to `null` in the `asException` code, so `initCause` throws an `IllegalArgumentException`.

### PR DESCRIPTION
**Pull request by rtadepalli**
Fixes #10900

### CODE REVIEW

**Comment by rtadepalli** [comment]
cc @ejona86

File: protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java [hunk 1]
```diff
@@ -176,6 +176,15 @@ public void fromThrowable_shouldReturnNullIfNoEmbeddedStatus() {
     assertNull(StatusProto.fromThrowable(nestedSe));
   }
 
+  @Test
+  public void toStatusExceptionWithMetadataAndCause_shouldCaptureCause() {
+    StatusException se = StatusProto.toStatusException(STATUS_PROTO, new Metadata(),
+        new RuntimeException("This is a test exception."));
```
Comment (ejona86) [inline protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:182]: If you save this exception, you can assert that it is the same identical exception instance.

```java
Throwable t = new RuntimeException("blah");
StatusException se = ...;

assertSame(t, se.getCause()); // assertEquals is also okay
```

File: protobuf/src/main/java/io/grpc/protobuf/StatusProto.java [hunk 2]
```diff
@@ -103,6 +103,25 @@ public static StatusException toStatusException(
...
 
+  /**
+   * Convert a {@link com.google.rpc.Status} instance to a {@link StatusException} with additional
+   * metadata and the root exception thrown.
+   *
+   * <p>The returned {@link StatusException} will wrap a {@link Status} whose code and description
+   * are set from the code and message in {@code statusProto}. {@code statusProto} will be
+   * serialized and added to {@code metadata}. {@code metadata} will be set as the metadata of the
+   * returned {@link StatusException}. The {@link Throwable} is the exception that is set as the
+   * {@code cause} of the returned {@link StatusException}.
```
Comment (ejona86) [inline protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:114]: Let's add a reminder here that cause is not transmitted over the wire.

Comment (rtadepalli) [reply protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:115]: Added it in the first paragraph.

File: protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java [hunk 3]
```diff
@@ -176,6 +176,15 @@ public void fromThrowable_shouldReturnNullIfNoEmbeddedStatus() {
     assertNull(StatusProto.fromThrowable(nestedSe));
   }
 
+  @Test
+  public void toStatusExceptionWithMetadataAndCause_shouldCaptureCause() {
+    RuntimeException exc = new RuntimeException("This is a test exception.");
+    StatusException se = StatusProto.toStatusException(STATUS_PROTO, new Metadata(), exc);
+
+    assertNotNull(se.getCause());
+    assertEquals(se.getCause(), exc);
```
Comment (panchenko) [inline protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:185]: `assertSame(exc, se.getCause())` - in JUnit the expected value is passed first.
`assertNotNull` above is redundant.

File: protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java [hunk 4]
```diff
@@ -176,6 +176,15 @@ public void fromThrowable_shouldReturnNullIfNoEmbeddedStatus() {
     assertNull(StatusProto.fromThrowable(nestedSe));
   }
 
+  @Test
+  public void toStatusExceptionWithMetadataAndCause_shouldCaptureCause() {
+    RuntimeException exc = new RuntimeException("This is a test exception.");
```
Comment (panchenko) [inline protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:181]: probably rename `exc` -> `cause`

File: protobuf/src/main/java/io/grpc/protobuf/StatusProto.java (hunk 2 above)
Comment (ejona86) [reply protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:115]: What does "needs to be user provided" mean?

File: protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java (hunk 4 above)
Comment (ejona86) [reply protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:181]: Yeah, `ex`, `exception`, `cause`, etc would be more common, but we can live with this.

**Review by ejona86 — APPROVED** [review]

**Review by larry-safran — APPROVED** [review]

**Pull request merged**